File: CallSim.py
import logging
import threading
import time
import random
import configparser

call_counter = 0
next_call = 0
allCalls = []
allWorkers = []
access_lock = threading.Lock()
stop_threads = 0

class call:
    def __init__ (self, callid, status):
        self.callid = callid
        self.status = status

class worker:
    def __init__ (self, status, calls_taken, efficiency):
        self.status
        self.calls_taken
        self.efficiency

def createCall(avgCalls, stdDev):
    global call_counter
    newCalls = avgCalls + random.randint(-stdDev, stdDev)
    for i in range (0, newCalls):
        c = call(call_counter, "on hold")
        allCalls.append(c)
        call_counter += 1

def processCall():
    global stop_threads
    while (stop_threads == 0):
        global access_lock
        # Critical selection of call
        access_lock.acquire()
        global next_call
        current_call = next_call
        next_call += 1
        access_lock.release()
        # If the call doesn't exits: wait
        while (len(allCalls)+1<current_call):
            #worker status: inactive
            print("waiting for call")
            time.sleep(1)
        print ("working on call: " + str(current_call) + "\n")
        allCalls[current_call].status = "processing"
        time.sleep(1)
        allCalls[current_call].status = "completed"


def print_calls ():
    for i in range (0, len(allCalls)):
        print("call " + str(allCalls[i].callid) + ", " + str(allCalls[i].status))
    
def main():
    # - - - - - - - - - - -  SETUP  - - - - - - - - - - - 
    #initiate logger
    print ("program started")
    random.seed()
    wthreads = []

    # input number of workers
    workers = int(input("Enter the number of workers: "))

    # read config file for time statistics
    config = configparser.ConfigParser()
    config.read("config.ini")
    avgCalls = int(config["Statistics"]["avgCalls"])
    stdCalls = int(config["Statistics"]["stdCalls"])
    avgCallTime = int(config["Statistics"]["avgCallTime"])
    contingency = int(config["Other Values"]["contingency"])

    if (avgCalls<stdCalls):
        print ("warning: there will be no calls in some cycles")
        

    # - - - - - - - - - - -  EXECUTION  - - - - - - - - - - -

    # Create a thread for each worker
    for i in range (0, workers):
        worker = threading.Thread(target=processCall)
        wthreads.append(worker)
        wthreads[i].start()

    # create calls in a cycle
    for i in range (0,12):
       createCall(avgCalls, stdCalls)
       time.sleep(1)

    # Work day is over
    global stop_threads
    stop_threads = 1
    
    # Join threads
    for i in range (0, workers):
        wthreads[i].join()

    print_calls()

File: reference/callsim.c
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include <pthread.h>
//#include <mpi.h>
   
#define max_rows 1000000
#define send_data_tag 2001
#define return_data_tag 2002
#define waiting 'w'
#define processing 'p'
#define complete 'c'

typedef struct {
	char node;
	char status;
	unsigned short wait;
} call_t;

typedef struct {
	unsigned short calls_taken;
	short efficiency;
} worker_t;

pthread_mutex_t access_lock;
int call_counter = 0;
int next_call = 0;
int call_pointer = 0;
int stop_threads = 0;

call_t allCalls[max_rows];
//call_t allCalls_node0[max_rows];

/** Create a number of calls based on statistics while there is room
 * \param avgCalls the average amount of calls in the time frame
 * \param stdDev the standard deviation of calls to add randomness
 */
void* createCall(){//int avgCalls, int stdDev) {
	int newCalls = 18;//avgCalls + rand()%stdDev-stdDev/2;
	for (int i=0; i<newCalls; i++){
		call_t nc = {'a', waiting, 0};
		if (call_pointer<max_rows){
			allCalls[call_pointer] = nc;
			call_pointer++;
		}
	}
}

/** Main thread cycle that updates calls. The work made by the
 * 	employees 
 */
void* processCall() {
	int current_call = 0;
	while (stop_threads == 0) {
		pthread_mutex_lock(&access_lock);
		current_call = next_call;
		next_call++;
		pthread_mutex_unlock(&access_lock);
		// If the call doesn't exist: wait
		while (current_call>=call_pointer) {
			printf("Waiting for a call\n");
			usleep(1000);
		}
		printf("Working on call %d\n", current_call);
		allCalls[current_call].status = processing;
		usleep(1000); // efficiency or statistics
		allCalls[current_call].status = complete;
	}
}

int main () {
	pthread_t allWorkers[1000];
	int max_workers = 20;
	int ret = 0;
	srand(time(0));

	createCall(); //so that the workers can immediately work ***(may not be realistic)
	
	// create threads
	for (int w = 0; w<max_workers; w++) {
		// append thread number to msg
		ret = pthread_create (&allWorkers[w], NULL, processCall, NULL);
		if (ret!=0) {printf("Error creating worker thread\n");}
	}

	// create calls in a cycle
	for (int i=0; i<49; i++){
		createCall();
		usleep(1000);
	}

	// work day is over
	stop_threads = 1;
	printf("+++ Total calls: %d +++\n", call_pointer-1);

	// join threads
	for (int w = 0; w<max_workers; w++) {
		pthread_join (allWorkers[w], NULL);
	}
}